pkg/controllers/provisioningv2/rke2/machinedrain/machinedrain.go

```diff
@@ -66,71 +67,109 @@ func (h *handler) k8sClient(machine *capi.Machine) (kubernetes.Interface, error)
 	return kubernetes.NewForConfig(restConfig)
 }
 
-func (h *handler) undrain(machine *capi.Machine) (*capi.Machine, error) {
+func (h *handler) undrain(machine *capi.Machine, drainData string) (*capi.Machine, error) {
 	if machine.Status.NodeRef == nil || machine.Status.NodeRef.Name == "" {
 		logrus.Debugf("unable to drain machine %s as there is no noderef", machine.Name)
 		return machine, nil
 	}
 
-	k8s, err := h.k8sClient(machine)
-	if err != nil {
-		return nil, err
-	}
-
-	node, err := k8s.CoreV1().Nodes().Get(h.ctx, machine.Status.NodeRef.Name, metav1.GetOptions{})
-	if err != nil {
-		return nil, err
+	var drainOpts rkev1.DrainOptions
+	if err := json.Unmarshal([]byte(drainData), &drainOpts); err != nil {
+		return machine, err
 	}
 
-	c := drain.NewCordonHelper(node)
-
-	if !c.UpdateIfRequired(false) {
-		if machine.Annotations[planner.UnCordonAnnotation] != "" {
-			machine = machine.DeepCopy()
-			delete(machine.Annotations, planner.UnCordonAnnotation)
+	if len(drainOpts.PostDrainHooks) > 0 {
+		if machine.Annotations[planner.PostDrainAnnotation] != drainData {
+			machine.Annotations[planner.PostDrainAnnotation] = drainData
 			return h.machines.Update(machine)
 		}
-		return machine, nil
+		for _, hook := range drainOpts.PostDrainHooks {
+			if hook.Annotation != "" && machine.Annotations[hook.Annotation] != drainData {
+				return machine, nil
+			}
+		}
 	}
 
-	err, patchErr := c.PatchOrReplaceWithContext(h.ctx, k8s, false)
+	helper, node, err := h.getHelper(machine, drainOpts)
 	if err != nil {
-		return machine, err
+		return nil, err
 	}
 
-	return machine, patchErr
-}
+	if err := drain.RunCordonOrUncordon(helper, node, false); err != nil {
+		return nil, err
+	}
 
-func (h *handler) drain(machine *capi.Machine, drainData []byte) (*capi.Machine, error) {
-	hash := planner.DrainHash(drainData)
-	if machine.Annotations[planner.DrainDoneAnnotation] == hash {
-		return machine, nil
+	if machine.Annotations[planner.UnCordonAnnotation] != "" {
```

Comment:
Same as above?
```suggestion
	if machine.Annotations[planner.UnCordonAnnotation] == drain {
```

Reply:
Also, a comment here would be nice. It looks like the uncordon has finished so we are resetting all the annotations so that the next drain happens as expected.

Reply:
This check is redundant so I've removed it.